yorkie/src/main/kotlin/dev/yorkie/document/crdt/CrdtRoot.kt

```diff
@@ -1,3 +1,188 @@
+     */
+    fun registerRemovedElement(element: CrdtElement) {
+        removedElementSetByCreatedAt.add(element.createdAt)
+    }
+
+    /**
+     * Registers the given [text] to the hast set.
+     */
+    fun registerTextWithGarbage(text: CrdtTextElement) {
+        textWithGarbageSetByCreatedAt.add(text.createdAt)
+    }
+
+    // NOTE(7hong13): the original comment from JS-SDK is as follows:
+    // `getGarbageLen` returns length of nodes which should garbage collection task
+    /**
+     * Returns length of nodes which can be garbage collected.
+     */
+    fun getGarbageLen(): Int {
+        var count = 0
+        removedElementSetByCreatedAt.forEach { createdAt ->
+            count++
+            val pair = elementPairMapByCreatedAt[createdAt] ?: return@forEach
+            if (pair.element is CrdtContainer) {
+                pair.element.getDescendants { _, _ ->
+                    count++
+                    false
+                }
+            }
+        }
+
+        textWithGarbageSetByCreatedAt.forEach { createdAt ->
+            val pair = elementPairMapByCreatedAt[createdAt] ?: return@forEach
+            val text = pair.element as CrdtTextElement
+            count += text.getRemovedNodesLen()
+        }
+
+        return count
+    }
+
+    /**
+     * Copies itself deeply.
+     */
+    fun deepCopy(): CrdtRoot {
+        return CrdtRoot(rootObject.deepCopy())
+    }
+
+    // NOTE(7hong13): parameter name is "ticket" in JS-SDK, but replaced it with "executedAt".
+    /**
+     * Deletes elements that were removed before [executedAt].
+     */
+    fun garbageCollect(executedAt: TimeTicket): Int {
+        var count = 0
+        removedElementSetByCreatedAt.forEach { createdAt ->
+            val pair = elementPairMapByCreatedAt[createdAt] ?: return@forEach
+            if (pair.element.isRemoved && pair.element.removedAt!! <= executedAt) {
+                pair.parent?.delete(pair.element)
+                count += garbageCollectInternal(pair.element)
+            }
+        }
+
+        textWithGarbageSetByCreatedAt.forEach { createdAt ->
+            val pair = elementPairMapByCreatedAt[createdAt] ?: return@forEach
+            val text = pair.element as CrdtTextElement
+
+            val removedNodeCount = text.deleteTextNodesWithGarbage(executedAt)
+            if (removedNodeCount > 0) {
+                textWithGarbageSetByCreatedAt.remove(text.createdAt)
+            }
+            count += removedNodeCount
+        }
+
+        return count
+    }
+
+    private fun garbageCollectInternal(element: CrdtElement): Int {
+        var count = 0
+        val callback = { elem: CrdtElement, _: CrdtContainer? ->
+            deregisterElement(elem)
+            count++
+            false
+        }
+        callback(element, null)
+        if (element is CrdtContainer) {
+            element.getDescendants(callback)
+        }
+        return count
+    }
+
+    private fun deregisterElement(element: CrdtElement) {
```

Comment:
it is public function in `yorkie-js-sdk`, but changed it as private since it is only used in `CrdtRoot.garbageCollectInternal` function.